internal/testrunner/runners/system/tester.go

```diff
@@ -1583,6 +1535,118 @@ func (r *tester) createServiceStateDir() error {
+	if config.WaitForDataTimeout > 0 {
+		waitForDataTimeout = config.WaitForDataTimeout
+	}
+
+	if config.Assert.HitCount > elasticsearchQuerySize {
+		return nil, fmt.Errorf("invalid value for assert.hit_count (%d): it must be lower of the maximum query size (%d)", config.Assert.HitCount, elasticsearchQuerySize)
+	}
+
+	if config.Assert.MinCount > elasticsearchQuerySize {
+		return nil, fmt.Errorf("invalid value for assert.min_count (%d): it must be lower of the maximum query size (%d)", config.Assert.MinCount, elasticsearchQuerySize)
+	}
+
+	// (TODO in future) Optionally exercise service to generate load.
+	logger.Debugf("checking for expected data in data stream (%s)...", waitForDataTimeout)
+	var hits *hits
+	oldHits := 0
+	foundFields := map[string]any{}
+	passed, waitErr := wait.UntilTrue(ctx, func(ctx context.Context) (bool, error) {
+		var err error
+		hits, err = r.getDocs(ctx, dataStream)
+		if err != nil {
+			return false, err
+		}
+
+		defer func() {
+			oldHits = hits.size()
+		}()
+
+		if r.checkFailureStore {
+			failureStore, err := r.getFailureStoreDocs(ctx, dataStream)
+			if err != nil {
+				return false, fmt.Errorf("failed to check failure store: %w", err)
+			}
+			if n := len(failureStore); n > 0 {
+				// Interrupt loop earlier if there are failures in the document store.
+				logger.Debugf("Found %d hits in the failure store for %s", len(failureStore), dataStream)
+				return true, nil
+			}
+		}
+
+		assertHitCount := func() bool {
+			if config.Assert.HitCount == 0 {
+				// not enabled
+				return true
+			}
+			if hits.size() < config.Assert.HitCount {
+				return false
+			}
+
+			ret := hits.size() == oldHits
+			if !ret {
+				time.Sleep(4 * time.Second)
+			}
+
+			return ret
+		}()
+
+		assertFieldsPresent := func() bool {
+			if len(config.Assert.FieldsPresent) == 0 {
+				// not enabled
+				return true
+			}
+			if hits.size() == 0 {
+				// At least there should be one document ingested
+				return false
+			}
+			for _, f := range config.Assert.FieldsPresent {
+				if _, found := foundFields[f]; found {
+					continue
+				}
+				found := false
+				for _, d := range hits.Fields {
+					if _, err := d.GetValue(f); err == nil {
+						found = true
+						break
+					}
+				}
+				if !found {
+					logger.Debugf("Not found field %q in hits", f)
+					return false
+				}
+				logger.Debugf("> Found field %q in hits", f)
+				foundFields[f] = struct{}{}
+			}
+			return true
+		}()
+
+		assertMinCount := func() bool {
+			if config.Assert.MinCount > 0 {
+				return hits.size() >= config.Assert.MinCount
+			}
+			// By default at least one document
+			return hits.size() > 0
+		}()
+
+		return assertFieldsPresent && assertMinCount && assertHitCount, nil
+	}, 1*time.Second, waitForDataTimeout)
+
+	if waitErr != nil {
+		return nil, waitErr
+	}
+
+	if !passed {
+		return nil, testrunner.ErrTestCaseFailed{Reason: fmt.Sprintf("could not find hits in %s data stream", dataStream)}
```

Comment:
Nit. Maybe we should rewrite this error. It can fail now for different reasons, even when hits are found.

Reply:
Something like `could not find the expected hits in %d data stream` ?

Reply:
Yes, something like this would be better.

The best might be to explicitly say what assertions are failing, but maybe this requires a number of changes on the current code, no need to do it by now.